Creating a JRibbonFrame wihout an application menu causes an NPE

greenwoodma opened the pull request:
I'm not sure if this is related to #78 or not, but if you create an instance of `JRibbonFrame` and don't set an app menu then this triggers an NPE in the latest 2.0.1 release of Flamingo as well as 2.5-SNAPSHOT on line 471

https://github.com/kirill-grouchnikov/radiance/blob/1c9325c62873c32444d493d667d45800d9621d1c/flamingo/src/main/java/org/pushingpixels/flamingo/internal/ui/ribbon/BasicRibbonUI.java#L471

The problem seems to be that the `applicationMenuButton` is only initialized if it is going to be shown (which makes sense)

https://github.com/kirill-grouchnikov/radiance/blob/1c9325c62873c32444d493d667d45800d9621d1c/flamingo/src/main/java/org/pushingpixels/flamingo/internal/ui/ribbon/BasicRibbonUI.java#L141

This PR fixes two places in `JRibbonFrame` where the `applicationMenuButton` was being accessed even though it was null. Testing seems to suggest that this fully fixes the NPE (or at least I can now create a `JRibbonFrame` without an application menu).

greenwoodma commented:
Yes that was deliberate as again if the app menu isn't enabled that causes an NPE and as far as I could tell the logic made more sense with the else removed.

kirill-grouchnikov commented:
![image](https://user-images.githubusercontent.com/52874/59152156-8e1cbf00-8a0c-11e9-8fe6-6740757cd9b6.png)

This doesn't seem to be right. In line 550 the application menu button is set to visible, then it gets its bounds, and then in line 558 it is marked to be non-visible. I haven't run this yet, but it looks like application menu button is never shown under any circumstances.

kirill-grouchnikov commented:
![image](https://user-images.githubusercontent.com/52874/59152163-cb814c80-8a0c-11e9-8620-54c27b7ef6ac.png)

This time with line numbers

greenwoodma commented:
You're right, I've messed something up there. The original code was definitely causing an NPE but your right my fix of removing the `else` is clearly a mistake. I didn't spot it as of course as I was testing with a use case where the button shouldn't be displayed *not that that's a great excuse).

I'll have another look some time tomorrow and see if I can figure out a better fix for that section.

greenwoodma commented:
A closer look and I see how I messed this up. When originally testing a fix I just commented out the original code, but when I cleaned things up for the pull request I messed up the issue around line 555. What I intended to do was to remove the entire else block. The reason being that you only fall into the else block if `ribbon.getApplicationMenuProjection()` returns null (i.e. the test on line 549 in your screenshot returns false). If `ribbon.getApplicationMenuProjection()` returns null then that means the `applicationMenuButton` is false which is why the original code then threw an NPE when calling `applicationMenuButton.setVisible(false);`. Removing the entire else block fixes all this simply as there is no need to make it invisible if it's never even been initialised.

I've updated the code in the PR so it should now make more sense than the original commit. Sorry for the confusion.

ultrasol commented:
Dear sir i have face problem in java swing flamingo substance jaspersoft ireport GUI print button show exception error need help.
